### PR TITLE
Jacobi Method

### PR DESCRIPTION
**Pull request by GM07**
This PR contains :

- Implementation of the Jacobi Algorithm
- Associated documentation
- Associated tests to validate the algorithm

### CODE REVIEW

File: src/jacobi.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#ifndef DIGITS_PRECISION
+    #define DIGITS_PRECISION 15
+#endif
+
+typedef double real_number;
+typedef unsigned int vec_size;
+
+#ifndef DIGITS_PRECISION
+    #define DIGITS_PRECISION 15
+#endif
```
Comment (bisson2000) [inline src/jacobi.h:17]: There are two defines for `DIGITS_PRECISION`. One should be removed
Also, do you think it should be a parameter passed to the function instead of being a define?

**Review by bisson2000 — CHANGES_REQUESTED** [review]

File: src/jacobi.c [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+
+
+/**
+ * Creates the identity matrix (1 on the diagonal and 0 on other coordinates)
+ * @param size size of the matrix
+ * @param output output matrix
+*/
+void jacobiCreateIdentityMatrix(vec_size size, real_number* output) {
+    for (vec_size i = 0; i < size * size; ++i) {
+        if (i / size == i % size) {
```
Comment (felix642) [inline src/jacobi.c:129]: Do you think we should create an inline method isDiagonal(index) since this code is used at multiple places?

File: src/jacobi.c [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+ * @param size size of the matrix (matrix must be square)
+ * @param outputMatrix matrix containing every eigenvector of the input matrix (same size as the input matrix)
+ * @param iterations number of iterations of the algorithm
+ * @param cyclic whether to use the cyclic jacobi method or the classical version
+*/
+void jacobi(real_number* inputMatrix, vec_size size, real_number* outputMatrix, vec_size iterations, vec_size cyclic) {
+
+    vec_size sizes3d[3] = {size, size, size};
+    vec_size sweepSize = size * (size - 1) / 2; // A sweep is defined as a n * (n - 1) / 2 jacobi rotations
+    real_number p[size * size]; // Rotation matrix
```
Comment (felix642) [inline src/jacobi.c:165]: If it's the rotation matrix, why is it called p? You could simply name it rotationMatrix and remove that comment no ?

File: src/jacobi.c [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+                indexToMinimize = (indexToMinimize + 1) % (sizeSquared);
+            }
+
+            maxRow = indexToMinimize / size;
+            maxCol = indexToMinimize % size;
+        } else {
+            jacobiMaxIndex(inputMatrix, size, &maxRow, &maxCol);
+        }
+
+        if (i < 3 * sweepSize) {
```
Comment (felix642) [inline src/jacobi.c:191]: magic number, 3?

File: src/jacobi.c [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            // In the first three sweeps, we keep the rotation matrix if |a_pq| > epsilon
+            real_number epsilon_sweep = 0.20 * (jacobiComputeOffDiagonalSum(inputMatrix, size)) / ((real_number) (sizeSquared));
```
Comment (felix642) [inline src/jacobi.c:193]: magic_number 0.2?

File: src/jacobi.c [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        } else {
+
+            real_number apq = real_abs(inputMatrix[coordToIndex(maxRow, maxCol, size)]);
+            real_number app = inputMatrix[coordToIndex(maxRow, maxRow, size)];
+            real_number aqq = inputMatrix[coordToIndex(maxCol, maxCol, size)];
+
+            // If |apq| << |app| and |apq << |aqq|, set the element to 0 and continue with other element
+            if (apq < pow(10, -(DIGITS_PRECISION + 2)) * app && apq < pow(10, -(DIGITS_PRECISION + 2) * aqq)) {
+                inputMatrix[coordToIndex(maxRow, maxCol, size)] = 0;
+                i--;
```
Comment (felix642) [inline src/jacobi.c:206]: --i; (yes, yes, I know it doesn't change anything but please :) )

File: src/jacobi.h [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+typedef double real_number;
+typedef unsigned int vec_size;
+
+#ifndef DIGITS_PRECISION
+    #define DIGITS_PRECISION 15
```
Comment (felix642) [inline src/jacobi.h:16]: Hmm, I would guess that, at this point, DIGIT_PRECISION will be defined... (Because it was defined 10 lines above)

File: src/jacobi.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+void jacobiMaxIndex(real_number* matrix, vec_size size, vec_size* maxX, vec_size* maxY);
+void jacobiMatrixMultiply(real_number* firstMatrix, real_number* secondMatrix, vec_size size[3], real_number* output, vec_size transposeFirstMatrix);
+void jacobiCreateRotationMatrix(real_number* input, vec_size row, vec_size col, real_number* output, vec_size size);
+void jacobi(real_number* inputMatrix, vec_size size, real_number* outputMatrix, vec_size iterations, vec_size cyclic);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
```
Comment (felix642) [inline src/jacobi.h:32]: This #endif as no matching #if

File: tests/test_jacobi.c [hunk 9]
```diff
@@ -0,0 +1,99 @@
+#include "../src/jacobi.h"
```
Comment (felix642) [inline tests/test_jacobi.c:1]: Relative imports will definitely break in the future. Please use #include <jacobi.h> and make sure that when compiling you use -I../src/ (Which should already be present in the makefile)

File: tests/test_jacobi.c [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+#include "stdio.h"
+
+#define PI 3.141592653589793238
+#define EPSILON 0.001
```
Comment (felix642) [inline tests/test_jacobi.c:5]: Should we make this parameter configurable ?

File: tests/test_jacobi.c [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+int testJacobi(real_number* input, real_number* expectedEigenValues, real_number* expectedEigenVectors, vec_size size, vec_size iterations, vec_size cyclic) {
+
+    vec_size squaredSize = size * size;
+
+    real_number mat1[squaredSize];
+    memcpy(mat1, input, sizeof(real_number) * squaredSize);
+
+    real_number output[squaredSize];
+    memset(output, 0, sizeof(real_number) * squaredSize);
```
Comment (felix642) [inline tests/test_jacobi.c:30]: Oh, that's interesting. We have to memset the output matrix. Is that written somewhere in the documentation? I don't remember seeing that?

File: Makefile [hunk 12]
```diff
@@ -8,23 +10,29 @@ CFLAGS += -I./src # included header files
...
+jacobi: ./$(TEST_FOLDER)/test_jacobi.c ./src/jacobi.c | build_folder
+	$(CC) $(CFLAGS) $^ -o $(BUILD_FOLDER)/test_$@.out $(LDLIBS)
 
 run_all_tests:
-	./test_linear_congruential_random_generator.out
-	./test_gauss_elimination.out
-	./test_poly_interpolation.out
+	./$(BUILD_FOLDER)/test_linear_congruential_random_generator.out
+	./$(BUILD_FOLDER)/test_gauss_elimination.out
+	./$(BUILD_FOLDER)/test_poly_interpolation.out
```
Comment (felix642) [inline Makefile:32]: You forgot to add jacobi to this section.

File: src/jacobi.c [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+            jacobiCreateRotationMatrix(inputMatrix, maxRow, maxCol, p, size);
+        }
+
+        jacobiMatrixMultiply(p, inputMatrix, sizes3d, aBuffer, 1);
+        jacobiMatrixMultiply(aBuffer, p, sizes3d, inputMatrix, 0);
+        jacobiMatrixMultiply(outputMatrix, p, sizes3d, vBuffer, 0);
+        memcpy(outputMatrix, vBuffer, sizeof(real_number) * sizeSquared);
+    }
+
+}
```
Comment (felix642) [inline src/jacobi.c:219]: No newline at the end of file ? 

Absolutely barbaric. If you're using vscode there's an option to automatically append newlines at the end, I think you should enable that ;)

**Review by felix642 — CHANGES_REQUESTED** [review]
@GM07 Good job! 

I've left a few comments that need to be addressed before merging but other than that, the code looks good.

File: src/jacobi.c [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+    real_number aBuffer[size * size];
+    vec_size maxRow, maxCol, indexToMinimize;
+    vec_size sizeSquared = size * size;
+
+    jacobiCreateIdentityMatrix(size, outputMatrix);
+
+    for (vec_size i = 0; i < iterations; ++i) {
+
+        if (cyclic) {
+            if (indexToMinimize == (sizeSquared - 2)) {
```
Comment (felix642) [inline src/jacobi.c:176]: This is UB, the variable indexToMinimize is defined at line 168, but never initialized here

**Comment by felix642** [comment]
@GM07, I tried running your test program on my machine and there seems to be a lot of UB in your code. If you run your program with Valgrind or ASAN you will see a lot of errors. 

I think we should fix those issues before merging

File: src/jacobi.c [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+    real_number c = 1.0 / (sqrt(1 + t * t));
+    real_number s = c * t;
+
+    vec_size cIndex1 = coordToIndex(row, row, size);
+    vec_size cIndex2 = coordToIndex(col, col, size);
+    vec_size sIndex1 = coordToIndex(row, col, size);
+    vec_size sIndex2 = coordToIndex(col, row, size);
+    vec_size sizeSquared = size * size;
+
+    for (vec_size i = 0; i < sizeSquared; ++i) {
```
Comment (nicholaslegrand102) [inline src/jacobi.c:106]: Would it be possible to make this into a switch?

File: src/jacobi.h (hunk 8 above)
Comment (nicholaslegrand102) [inline src/jacobi.h:32]: Missing newline

File: tests/test_jacobi.c [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        0.541774,   -0.707107,  0.454401 ,
+        0.642621,   0.000000 ,  -0.766185,
+        0.541774,   0.707107 ,  0.454401 ,
+    };
+
+    fail |= testJacobi(matTestJacobi, expectedEigenValues, expectedEigenVectors, 3, 10, 1);
+    fail |= testJacobi(matTestJacobi, expectedEigenValues, expectedEigenVectors, 3, 10, 0);
+
+    return fail;
+}
```
Comment (nicholaslegrand102) [inline tests/test_jacobi.c:99]: Missing newline

File: src/jacobi.h (hunk 1 above)
Comment (GM07) [reply src/jacobi.h:17]: I don't think it should be a parameter passed to the function, because this will not vary between multiple calls of the function so it would be redundant for the user to always put it when calling the function

File: src/jacobi.c (hunk 2 above)
Comment (GM07) [reply src/jacobi.c:129]: Yes good point !

File: src/jacobi.c (hunk 4 above)
Comment (GM07) [reply src/jacobi.c:191]: This is an optimization proposed in Numerical Recipes

File: tests/test_jacobi.c (hunk 11 above)
Comment (GM07) [reply tests/test_jacobi.c:30]: we don't have too, I removed it

File: src/jacobi.c (hunk 15 above)
Comment (GM07) [reply src/jacobi.c:106]: I cannot make this into a switch, because the values compared in a switch statement must be known at compile time. In my case, they are not

File: src/jacobi.c (hunk 4 above)
Comment (felix642) [reply src/jacobi.c:189]: Ok, do we know why we are using 3 here? I think we should either add a comment explaining this number or replace 3 with a constant equal to 3

Comment (GM07) [reply src/jacobi.c:189]: I have a comment right under the condition, but I can add the fact that this number was proposed in the book Numerical Recipes

Comment (felix642) [reply src/jacobi.c:189]: The comment under it explains what we do during the first 3 sweeps, but not why we chose that number. If we don't know precisely why we decided to use 3 we should add a comment explaining that it was proposed in the book.

File: src/jacobi.c [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+    real_number maxElement = matrix[1]; // Cannot be first element since it's on the diagonal
+    vec_size currentMaxRow = 0;
+    vec_size currentMaxCol = 1;
+
+    // Multiplication could overflow if vec_size is not defined properly
+    vec_size sizeSquared = size * size;
+
+    // Iteration starts at 2 since we assume that the first element off diagonal is the max at that point
+    for (vec_size i = 2; i < sizeSquared; ++i) {
+        vec_size currentRow = i / size;
```
Comment (jmsellier) [inline src/jacobi.c:32]: try to remove division (use two ifs instead)

File: src/jacobi.c [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+ */
+void jacobiMatrixMultiply(real_number* firstMatrix, real_number* secondMatrix, vec_size size[3], real_number* output, vec_size transposeFirstMatrix) {
+    
+    vec_size m = size[0];
+    vec_size n = size[1];
+    vec_size p = size[2];
+    for (vec_size i = 0; i < m; ++i) {
+        for (vec_size j = 0; j < p; ++j) {
+            real_number sum = 0;
+            for (vec_size k = 0; k < n; ++k) {
```
Comment (jmsellier) [inline src/jacobi.c:64]: register keyword

File: src/jacobi.c [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+ * @param transposeFirstMatrix whether to transpose the first matrix when multiplying the matrices
+ */
+void jacobiMatrixMultiply(real_number* firstMatrix, real_number* secondMatrix, vec_size size[3], real_number* output, vec_size transposeFirstMatrix) {
+    
+    vec_size m = size[0];
+    vec_size n = size[1];
+    vec_size p = size[2];
+    for (vec_size i = 0; i < m; ++i) {
+        for (vec_size j = 0; j < p; ++j) {
+            real_number sum = 0;
```
Comment (jmsellier) [inline src/jacobi.c:63]: 0.0

File: src/jacobi.c [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+*/
+void jacobi(real_number* inputMatrix, vec_size size, real_number* outputMatrix, vec_size iterations, vec_size cyclic) {
+
+    if (size == 1) {
+        memcpy(outputMatrix, inputMatrix, sizeof(real_number));
+        return;
+    }
+
+    vec_size sizes3d[3] = {size, size, size};
+    vec_size sweepSize = size * (size - 1) / 2; // A sweep is defined as a n * (n - 1) / 2 jacobi rotations
```
Comment (jmsellier) [inline src/jacobi.c:162]: shift

File: src/jacobi.c [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+                jacobiCreateRotationMatrix(inputMatrix, maxRow, maxCol, rotationMatrix, size);
+            }
+        } else {
+
+            real_number apq = fabs(inputMatrix[coordToIndex(maxRow, maxCol, size)]);
+            real_number app = inputMatrix[coordToIndex(maxRow, maxRow, size)];
+            real_number aqq = inputMatrix[coordToIndex(maxCol, maxCol, size)];
+
+            // If |apq| << |app| and |apq << |aqq|, set the element to 0 and continue with other element
+            if (apq < pow(10, -(DIGITS_PRECISION + 2)) * app && apq < pow(10, -(DIGITS_PRECISION + 2) * aqq)) {
```
Comment (jmsellier) [inline src/jacobi.c:202]: fastpow

**Comment by GM07** [comment]
> @felix642, These commits fix some issues with the algorithm too. Also, the algorithm should work on an arduino and the test zero does not verify that. Since the prototype is done and is a good test to see if it works on the arduino, don't you think it should be included ? As for the review process, the python program added only takes care of sending the data from the python program to the arduino and retrieve back the eigenvectors and eigenvalues and multiplies a few matrices to reconstruct the compressed image. Plus, most of the code has already been reviewed. We can talk about it if you want !

**Comment by felix642** [comment]
Well, I need those changes to start my prototype since it requires a working version of Jacobi. But I assume that I can rebase my changes on top of what you did and start from there. 

I'll re-request a review from everybody since you changed a bunch of things and try to review your code later tonight.

File: prototype/pca_compression/compress.py [hunk 22]
```diff
@@ -0,0 +1,75 @@
+import serial
```
Comment (nicholaslegrand102) [inline prototype/pca_compression/compress.py:1]: Is it ok for us to be dependant on pySerial

File: prototype/pca_compression/compress.py [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+import numpy as np
+from PIL import Image
+import re
+import matplotlib.pyplot as plt
+
+def encode_arduino_input(matrix):
```
Comment (nicholaslegrand102) [inline prototype/pca_compression/compress.py:7]: Do we want comments on our prototypes as well?

File: prototype/pca_compression/compress.py (hunk 22 above)
Comment (GM07) [reply prototype/pca_compression/compress.py:1]: The prototype that was already present in the library (mnist) depends on tensorflow, panda and sklearn. Therefore, I assumed it was alright to depend on it, but we could give a requirements.txt in the prototype folder to allow users to directly run every prototypes.

File: prototype/pca_compression/compress.py (hunk 23 above)
Comment (GM07) [reply prototype/pca_compression/compress.py:7]: What do you mean comments ? Like if we review the prototypes too or if we comment what functions do in prototypes

File: prototype/pca_compression/pca_compression.ino [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+
+
+void process() {
+  INTEGER sizes[3] = {size, size, size};
+
+  REAL_NUMBER* out = new REAL_NUMBER[totalSize];
+
+  jacobi(data, size, out, -1, 0);
+
+  Serial.print("<");
```
Comment (bisson2000) [inline prototype/pca_compression/pca_compression.ino:72]: It would be nice to add comments explaining how transmission and reception works.

File: prototype/pca_compression/pca_compression.ino [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+INTEGER totalSize = 0;
+REAL_NUMBER data[MATRIX_SIZE * MATRIX_SIZE];
+
+void read() {
+  if (Serial.available()) {
+    INTEGER i = 0;
+    String input = Serial.readString();
+    if (input.length() > 0) {
+      // parse input string to extract size and data
+      char* str = new char[input.length() + 1];
```
Comment (felix642) [inline prototype/pca_compression/pca_compression.ino:34]: I think this variable can live on the stack ?

File: prototype/pca_compression/pca_compression.ino [hunk 26]
```diff
@@ -0,0 +1,101 @@
+#define REAL_NUMBER float
+#define INTEGER uint8_t
+#define DIGITS_PRECISION 3
+#define MATRIX_SIZE 8
```
Comment (felix642) [inline prototype/pca_compression/pca_compression.ino:4]: Don't we receive the matrix size or the number of elements through the serial port?

File: src/jacobi.c [hunk 27]
```diff
@@ -0,0 +1,226 @@
...
+
+/**
+ * Converts a 2d coordinate in a square matrix to an index in a 2d array
+*/
+static inline vec_size coordToIndex(vec_size row, vec_size col, vec_size size) {
+    return row * size + col;
+}
+
+static inline vec_size isDiagonal(vec_size index, vec_size size) {
+    return (vec_size) ((index / size) == (index % size));
```
Comment (felix642) [inline src/jacobi.c:11]: The cast should be implicit here.

File: src/jacobi.c [hunk 28]
```diff
@@ -0,0 +1,226 @@
...
+ * Computes the jacobi method to find the eigenvalues and eigenvectors of the input matrix
+ * @param inputMatrix matrix whose eigenvalues and eigenvectors we are looking for (will contain the eigenvalues after this function has been called)
+ * @param size size of the matrix (matrix must be square)
+ * @param outputMatrix matrix containing every eigenvector of the input matrix (same size as the input matrix)
+ * @param max_iterations max number of iterations the algorithm can do (set to -1 to allow any amount of iterations)
+ * @param cyclic whether to use the cyclic jacobi method or the classical version
+*/
+void jacobi(real_number* inputMatrix, vec_size size, real_number* outputMatrix, vec_size max_iterations, vec_size cyclic) {
+
+    if (size == 1) {
```
Comment (felix642) [inline src/jacobi.c:160]: Do we really need to handle the case where the size is one? This looks like wasted cycles to me. What do you think ?

File: src/jacobi.c [hunk 29]
```diff
@@ -0,0 +1,226 @@
...
+    vec_size sweepSize = (size * (size - 1) >> 1); // A sweep is defined as a n * (n - 1) / 2 jacobi rotations
+    real_number rotationMatrix[size * size];
+    real_number vBuffer[size * size];
+    real_number aBuffer[size * size];
+    vec_size maxRow, maxCol, indexToMinimize = 0;
+    vec_size sizeSquared = size * size;
+
+    jacobiCreateIdentityMatrix(size, outputMatrix);
+    real_number currentOffDiagonalSum = jacobiComputeOffDiagonalSum(inputMatrix, size);
+    vec_size allowInfiniteIterations = (vec_size) (max_iterations == -1);
```
Comment (felix642) [inline src/jacobi.c:175]: Conversion should be implicit here

File: src/jacobi.h [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+
+typedef REAL_NUMBER real_number;
+typedef INTEGER vec_size;
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+void (*serialLog) (char*);
```
Comment (felix642) [inline src/jacobi.h:31]: I doubt those are needed in the header file.

File: tests/test_jacobi.c [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+
+    vec_size maxRow;
+    vec_size maxCol;
+    jacobiMaxIndex(input, size, &maxRow, &maxCol);
+
+    if (maxRow == expectedRow && maxCol == expectedCol) {
+        printf("Success : Test maximum index\n");
+        return 0;
+    } else {
+        printf("Fail : Test maximum index. Row should be 1 and column should be 0, but row is %d and column is %d\n", maxRow, maxCol);
```
Comment (felix642) [inline tests/test_jacobi.c:15]: Message should say Row should be expectedRow and column should be...

**Review by felix642 — CHANGES_REQUESTED** [review]

File: src/jacobi.c (hunk 28 above)
Comment (GM07) [reply src/jacobi.c:160]: True

**Review by felix642 — APPROVED** [review]
LGTM

**Review by bisson2000 — APPROVED** [review]
lgtm
